Framework/PythonInterface/plugins/algorithms/MaskAngle.py

```diff
@@ -42,61 +44,56 @@ def PyInit(self):
         self.declareProperty(mantid.kernel.IntArrayProperty(name="MaskedDetectors", direction=mantid.kernel.Direction.Output),
-                             doc="List of detector masked, with scatterin angles between MinAngle and MaxAngle")
+                             doc="List of detector masked, with scattering angles between MinAngle and MaxAngle")
 
     def validateInputs(self):
         issues = dict()
-
         ws = self.getProperty("Workspace").value
-
-        try:
-            if type(ws).__name__ == "WorkspaceGroup":
-                for w in ws:
-                    w.getInstrument().getSource().getPos()
-            else:
-                ws.getInstrument().getSource().getPos()
-        except (RuntimeError, ValueError, AttributeError, TypeError):
+        hasInstrument = True
+        if type(ws).__name__ == "WorkspaceGroup" and len(ws) > 0:
+            for item in ws:
+                hasInstrument = hasInstrument and len(item.componentInfo()) > 0
+        else:
+            hasInstrument = len(ws.componentInfo()) > 0
+        if not hasInstrument:
             issues["Workspace"] = "Workspace must have an associated instrument."
-
         return issues
 
+    def _get_phi(self, det_index, detector_info):
+        pos = detector_info.position(det_index)
+        # see Detector.cpp for phi definition
+        return math.fabs(math.atan2(pos.Y(), pos.X()))
+
     def PyExec(self):
         ws = self.getProperty("Workspace").value
         ttmin = numpy.radians(self.getProperty("MinAngle").value)
         ttmax = numpy.radians(self.getProperty("MaxAngle").value)
         if ttmin > ttmax :
             raise ValueError("MinAngle > MaxAngle, please check angle range for masking")
 
-        angle = self.getProperty('Angle').value
-
-        detlist=[]
+        angle_phi = self.getProperty('Angle').value == 'Phi'
 
         numspec = ws.getNumberHistograms()
         spectrumInfo = ws.spectrumInfo()
-
-        if angle == 'Phi':
-            for i in range(numspec):
-                if not spectrumInfo.isMonitor(i):
-                    det = ws.getDetector(i)
-                    phi=abs(det.getPhi())
-                    if phi>= ttmin and phi<= ttmax:
-                        detlist.append(det.getID())
-        else:
-            source=ws.getInstrument().getSource().getPos()
-            sample=ws.getInstrument().getSample().getPos()
-            beam = sample-source
-            for i in range(numspec):
-                if not spectrumInfo.isMonitor(i):
-                    det = ws.getDetector(i)
-                    tt=det.getTwoTheta(sample,beam)
-                    if tt>= ttmin and tt<= ttmax:
-                        detlist.append(det.getID())
-
-        if len(detlist)> 0:
-            mantid.simpleapi.MaskDetectors(Workspace=ws,DetectorList=detlist)
-        else:
+        detectorInfo = ws.detectorInfo()
+        det_ids = detectorInfo.detectorIDs()
+        masked_ids = list()
+        for i in range(numspec):
```

Comment:
Can't you use the `SpectrumInfo` iterator here?

Reply:
Wasn't available when I started writing this, yes will update